### ISSUE TITLE
Adding edges might violate Liskov 

### ISSUE THREAD

**Issue by cmwedin** [cmwedin/CustomGraphs#16]
an abstract graph can have an edge added to it and will only throw an exception if that edge is already attached to another graph, if one of the new ID's isnt on the graph it will make a new node with that ID. It can also have an edge added between two of its child nodes without throwing an exception. As part of the definition of a tree adding any edge to one would create a cycle, making it no longer a valid tree. 

Hence, adding an edge between two existing nodes in a tree would cause an exception to be thrown. I could just make this case silently fail rather then throw an exception but that seems like an unideal solution

**Comment by cmwedin** [comment]
### Liskov complaint AddEdge() practices 

- All methods for adding edges should be abstract in AbstractGraph even if it means repeating code
- These methods should be limited to what follows: 
```
abstract void AddNewEdge(int id1, int id2, [float weight = 1])
abstract void AddNewEdge(Node node1, Node node2, [float weight = 1])
abstract void AddEdge(AbstractEdge edge) 
```
- The methods implementation should adhere to the following standard:
  - they should not throw exceptions
  - AddNewEdge will create a new node if one of the id's isnt in the graph.
    - if node1 and node 2 both have parents and they are different it will fail
    - if one or both is null (aka they are new orphan nodes) it will add them to the graph  
  - failure to add can occur for various reasons depending on the type of graph. (e.g. node1 and node2 have different parents, adding the edge would make a tree cyclic). When this happens the method should return early and print a warning in the console

I am still not entirely confident this approach is Liskov compliant. Requiring the edge in AddEdge(AbstractEdge edge) to be an undirected edge in one case and directed edge in the other seems like textbook precondition strengthening, but since it would be strength-ed in the first non-abstract implementation of the method I'm willing to accept it?

**Comment by cmwedin** [comment]
The final implementation of this was basically what I suggested above with slight different names (the distinction between adding an edge an adding a new edge is ambiguous when i can pass a new edge into the argument of AddEdge(edge)). The final name TryAddEdge for all three functions is also more explicit that adding the edge is not guaranteed by the method but rather that the method will attempt to and tell you if it  failed.

While the largest change in this solution is more clear semantics I more confident in its compliance with Liskov. If B is a subtype of A and A contains a method bool TryAction, it seems reasonable that if an instance of type A was replaced by type B the method would behave differently as well. Returning false is explicitly a possible result of the method and not an exception, and requiring edges to be a specific type before adding them is a direct change to the behavior of the method rather then a stronger precondition, as in all types off graphs TryAddEdge(Edge) will accept any kind of edge, even if it returns false immediately.
